File: 0036-valid-sudoku/0036-valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        HashSet<Character> colSet = new HashSet<Character>();
        HashSet<Character> rowSet = new HashSet<Character>();
        HashMap<String, HashSet<Character>> gridMap = new HashMap<>();
        for(int row = 0; row < 9; row++){
            for(int col = 0; col < 9; col++){
                char chRow = board[row][col];
                char chCol = board[col][row];
                if(chRow!='.' && rowSet.contains(chRow)){
                    return false;
                }
                else{
                    rowSet.add(chRow);
                }
                if(chCol!='.' && colSet.contains(chCol)){
                    return false;
                }
                else{
                    colSet.add(chCol);
                }
                String gridStr = row/3 + "." + col/3;
                if(gridMap.containsKey(gridStr)){
                    HashSet<Character> gridSet = gridMap.get(gridStr);
                    if(chRow!='.' && gridSet.contains(chRow)){
                        return false;
                    }
                    else{
                        gridSet.add(chRow);
                        gridMap.put(gridStr, gridSet);
                    }
                }
                else{
                    if(chRow!='.'){
                        gridMap.put(gridStr, new HashSet<Character>(Collections.singleton(chRow)));
                    }
                }
            }
            colSet.clear();
            rowSet.clear();
        }
        return true;
    }
}

File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rowSet, colSet, gridSet = defaultdict(set), defaultdict(set), defaultdict(set)
        for r in range(9):
            for c in range(9):
                value = board[r][c]
                if value == ".":
                    continue
                if value in rowSet[r] or value in colSet[c] or value in gridSet[(r//3, c//3)]:
                    return False
                else:
                    rowSet[r].add(value)
                    colSet[c].add(value)
                    gridSet[(r//3, c//3)].add(value)
        return True 